### ISSUE TITLE
FreezingException not caught when switching context on native

### ISSUE THREAD

**Issue by erikdewit87** [Kotlin/kotlinx.coroutines#2064]
Given: Object with state marked with `ensureNeverFrozen()` to detect accidental freezes.
When: `withContext(IO)` accidentially freezes `this`
Then: Expect `FreezingException` to be thrown and caught when switching context
Bug: `FreezingException` is thrown but not properly caught

gradle: gradle-5.6.2
kotlin = 1.3.72
coroutines = 1.3.5-native-mt
iOS simulator: iPhone 11 pro max running iOS 13.1

Slack discussion:
https://kotlinlang.slack.com/archives/C3SGXARS6/p1590498022256500

Example:
```kotlin
interface TestView {
    fun setContent(content: List<String>)
}

class TestInteractor {
    suspend fun fetch(): List<Int> {
        // simulate network / db interaction
        delay(1000L)
        return (0..100).toList()
    }
}

class TestPresenter(
    val view: TestView,
    val interactor: TestInteractor
) : CoroutineScope {

    val cache = mutableListOf<String>()
    val job = SupervisorJob()
    val exceptionHandler = CoroutineExceptionHandler { _, throwable ->
        println("Caught with CoroutineExceptionHandler:")
        throwable.printStackTrace()
    }

    @InternalCoroutinesApi
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job + exceptionHandler

    init {
        ensureNeverFrozen()
    }

    fun load() {
        launch {
            println("Start fetching")
            val contentFromNetwork = withContext(Dispatchers.Default) {
                interactor.fetch() // freeze this on purpose to trigger FreezeException
            }

            println(
                "this.isFrozen:${this@TestPresenter.isFrozen} " +
                "interactor.isFrozen:${interactor2.isFrozen} " +
                "cache.isFrozen:${cache.isFrozen}"
            )

            println("Processing")
            val content = contentFromNetwork.map { it.toString() }

            println("Updating local cache")
            cache.addAll(content)

            println("Updating view")
            view.setContent(cache.toList())
        }
    }
}
```

Output:
``` 
Start fetching
```

While fetching it fails silently. `exceptionHandler` does not report any exception making it hard to determine what goes wrong.

Wrapping `withContext` inside try/catch shows a `FreezingException` is thrown:
```kotlin
val contentFromNetwork = try {
    withContext(Dispatchers.Default) {
        interactor.fetch() // freeze this on purpose to trigger FreezeException
    }
} catch (throwable: Throwable) {
    println("Caught with try/catch:")
    throwable.printStackTrace()
    emptyList()
}
```

Output:
```
Start fetching
Start fetching
Caught with try/catch:
kotlin.native.concurrent.FreezingException: freezing of kotlinx.coroutines.$startCoroutine$lambda-0$FUNCTION_REFERENCE$30@2ebba58 has failed, first blocker is com.theothersonline.onecrowd.common.presentation.content.TestPresenter@35c3988
        at 0   onecrowd                            0x00000001074244e7 kfun:kotlin.Throwable.<init>(kotlin.String?)kotlin.Throwable + 87
        at 1   onecrowd                            0x00000001074172b5 kfun:kotlin.Exception.<init>(kotlin.String?)kotlin.Exception + 85
        at 2   onecrowd                            0x00000001074174d5 kfun:kotlin.RuntimeException.<init>(kotlin.String?)kotlin.RuntimeException + 85
        at 3   onecrowd                            0x000000010746bfa1 kfun:kotlin.native.concurrent.FreezingException.<init>(kotlin.Any;kotlin.Any)kotlin.native.concurrent.FreezingException + 641
        at 4   onecrowd                            0x000000010746eefe ThrowFreezingException + 222
        at 5   onecrowd                            0x000000010784c002 FreezeSubgraph + 2642
        at 6   onecrowd                            0x000000010788021b Kotlin_Worker_freezeInternal + 27
        at 7   onecrowd                            0x000000010746c0bb kfun:kotlin.native.concurrent.freeze@T.(){0<kotlin.Any?>}Generic + 59
        at 8   onecrowd                            0x0000000107c4c748 kfun:kotlinx.coroutines.execute$kotlinx-coroutines-core@kotlin.native.concurrent.Worker.(kotlin.Function0<kotlin.Unit>) + 152
        at 9   onecrowd                            0x0000000107c4cae2 kfun:kotlinx.coroutines.WorkerThread.execute(kotlin.Function0<kotlin.Unit>) + 98
        at 10  onecrowd                            0x0000000107c44903 kfun:kotlinx.coroutines.startCoroutine$kotlinx-coroutines-core(kotlinx.coroutines.CoroutineStart;kotlinx.coroutines.AbstractCoroutine<>T>;R;kotlin.coroutines.SuspendFunction1<R,T>){0<kotlin.Any?>;1<kotlin.Any?>}Generic + 1411
        at 11  onecrowd                            0x0000000107aa6be0 kfun:kotlinx.coroutines.withContext(kotlin.coroutines.CoroutineContext;kotlin.coroutines.SuspendFunction1<kotlinx.coroutines.CoroutineScope,T>){0<>kotlin.Any?>}Generic + 1328
        at 12  onecrowd                            0x00000001070f8a03 kfun:com.theothersonline.onecrowd.common.presentation.content.TestPresenter.$loadContent$lambda-1COROUTINE$23.invokeSuspend#internal + 2883
        at 13  onecrowd                            0x0000000107450818 kfun:kotlin.coroutines.native.internal.BaseContinuationImpl.resumeWith(kotlin.Result<kotlin.Any?>) + 712
        at 14  onecrowd                            0x0000000107c080c3 kfun:kotlinx.coroutines.DispatchedTask.run() + 2627
        at 15  onecrowd                            0x0000000107c5a385 kfun:kotlinx.coroutines.DarwinMainDispatcher.dispatch$lambda-0#internal + 85
        at 16  onecrowd                            0x0000000107c5a6fb kfun:kotlinx.coroutines.DarwinMainDispatcher.$dispatch$lambda-0$FUNCTION_REFERENCE$40.invoke#internal + 59
        at 17  onecrowd                            0x0000000107c5a75b kfun:kotlinx.coroutines.DarwinMainDispatcher.$dispatch$lambda-0$FUNCTION_REFERENCE$40.$<bridge-UNN>invoke()#internal + 59
        at 18  onecrowd                            0x0000000107c5b814 _6b6f746c696e782d636f726f7574696e65732d636f7265_knbridge8 + 180
        at 19  libdispatch.dylib                   0x00000001096eedf8 _dispatch_call_block_and_release + 12
        at 20  libdispatch.dylib                   0x00000001096efd6c _dispatch_client_callout + 8
        at 21  libdispatch.dylib                   0x00000001096fde24 _dispatch_main_queue_callback_4CF + 1500
        at 22  CoreFoundation                      0x00007fff23b0ce49 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
        at 23  CoreFoundation                      0x00007fff23b07aa9 __CFRunLoopRun + 2329
        at 24  CoreFoundation                      0x00007fff23b06e66 CFRunLoopRunSpecific + 438
        at 25  GraphicsServices                    0x00007fff38346bb0 GSEventRunModal + 65
        at 26  UIKitCore                           0x00007fff47578dd0 UIApplicationMain + 1621
        at 27  onecrowd                            0x0000000107096d1b main + 75 (/Users/<me>/Git/onecrowd/onecrowd-app/ios/onecrowd/onecrowd/AppDelegate/AppDelegate.swift:13:7)
        at 28  libdyld.dylib                       0x00007fff516ecd29 start + 1
        at 29  ???                                 0x0000000000000001 0x0 + 1
this.isFrozen:false interactor.isFrozen:false cache.isFrozen:false
Processing: 0
Updating local cache
Updating view
```

Nothing got frozen and an exception is thrown so `ensureNeverFrozen()` did its job. Notice the exception is caught locally and not by the `exceptionHandler`. Rethrowing inside the catch still fails silently and does not reach the scope's `exceptionHandler`

To complete the example here is the output when `ensureNeverFrozen()` is removed from the initial example. This crashes with an `InvalidMutabilityException` when updating the 
```
Start fetching
this.isFrozen:true interactor.isFrozen:true cache.isFrozen:true
Processing: 100
Updating local cache
kotlin.native.concurrent.InvalidMutabilityException: mutation attempt of frozen kotlin.collections.ArrayList@9064f8
       at 0   onecrowd                            0x0000000101455517 kfun:kotlin.Throwable.<init>(kotlin.String?)kotlin.Throwable + 87
       at 1   onecrowd                            0x00000001014482e5 kfun:kotlin.Exception.<init>(kotlin.String?)kotlin.Exception + 85
       at 2   onecrowd                            0x0000000101448505 kfun:kotlin.RuntimeException.<init>(kotlin.String?)kotlin.RuntimeException + 85
       at 3   onecrowd                            0x000000010149d095 kfun:kotlin.native.concurrent.InvalidMutabilityException.<init>(kotlin.String)kotlin.native.concurrent.InvalidMutabilityException + 85
       at 4   onecrowd                            0x00000001014a0218 ThrowInvalidMutabilityException + 680
       at 5   onecrowd                            0x0000000101892588 MutationCheck + 104
       at 6   onecrowd                            0x0000000101462630 kfun:kotlin.collections.ArrayList.<set-array>#internal + 96
       at 7   onecrowd                            0x0000000101464b41 kfun:kotlin.collections.ArrayList.ensureCapacity(kotlin.Int) + 593
       at 8   onecrowd                            0x000000010146533f kfun:kotlin.collections.ArrayList.ensureExtraCapacity#internal + 95
       at 9   onecrowd                            0x0000000101465bdb kfun:kotlin.collections.ArrayList.insertAtInternal#internal + 267
       at 10  onecrowd                            0x0000000101466327 kfun:kotlin.collections.ArrayList.addAllInternal#internal + 727
       at 11  onecrowd                            0x0000000101463ee0 kfun:kotlin.collections.ArrayList.addAll(kotlin.collections.Collection<E>)kotlin.Boolean + 192
       at 12  onecrowd                            0x000000010112a405 kfun:com.theothersonline.onecrowd.common.presentation.content.TestPresenter.$loadContent$lambda-1COROUTINE$23.invokeSuspend#internal + 5093
       at 13  onecrowd                            0x0000000101481848 kfun:kotlin.coroutines.native.internal.BaseContinuationImpl.resumeWith(kotlin.Result<kotlin.Any?>) + 712
       at 14  onecrowd                            0x0000000101c390f3 kfun:kotlinx.coroutines.DispatchedTask.run() + 2627
       at 15  onecrowd                            0x0000000101c8b3b5 kfun:kotlinx.coroutines.DarwinMainDispatcher.dispatch$lambda-0#internal + 85
       at 16  onecrowd                            0x0000000101c8b72b kfun:kotlinx.coroutines.DarwinMainDispatcher.$dispatch$lambda-0$FUNCTION_REFERENCE$40.invoke#internal + 59
       at 17  onecrowd                            0x0000000101c8b78b kfun:kotlinx.coroutines.DarwinMainDispatcher.$dispatch$lambda-0$FUNCTION_REFERENCE$40.$<bridge-UNN>invoke()#internal + 59
       at 18  onecrowd                            0x0000000101c8c844 _6b6f746c696e782d636f726f7574696e65732d636f7265_knbridge8 + 180
       at 19  libdispatch.dylib                   0x000000010371fdf8 _dispatch_call_block_and_release + 12
       at 20  libdispatch.dylib                   0x0000000103720d6c _dispatch_client_callout + 8
       at 21  libdispatch.dylib                   0x000000010372ee24 _dispatch_main_queue_callback_4CF + 1500
       at 22  CoreFoundation                      0x00007fff23b0ce49 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
       at 23  CoreFoundation                      0x00007fff23b07aa9 __CFRunLoopRun + 2329
       at 24  CoreFoundation                      0x00007fff23b06e66 CFRunLoopRunSpecific + 438
       at 25  GraphicsServices                    0x00007fff38346bb0 GSEventRunModal + 65
       at 26  UIKitCore                           0x00007fff47578dd0 UIApplicationMain + 1621
       at 27  onecrowd                            0x00000001010c7e8b main + 75 (/Users/erikdewit/Programming/Git/onecrowd/onecrowd-app/ios/onecrowd/onecrowd/AppDelegate/AppDelegate.swift:13:7)
       at 28  libdyld.dylib                       0x00007fff516ecd29 start + 1
       at 29  ???                                 0x0000000000000001 0x0 + 1
```

**Comment by elizarov** [comment]
`FreezingException` should be thrown locally by the design of `withContext`. The fact that it hangs after that is a bug. I'm fixing it in the upcoming `native-mt` release. Watch #462 for details.

**Comment by qwwdfsad** [comment]
Fixed in `1.3.9-native-mt-2`
